transport_plugins/bled112/iotile_transport_bled112/bled112.py

```diff
@@ -124,6 +129,15 @@ def find_bled112_devices(cls):
 
         return found_devs
 
+    def get_scan_stats(self):
+        return self._scan_event_count, self._v1_scan_count, self._v2_scan_count
+
+    def reset_scan_stats(self):
+        self._scan_event_count = 0
+        self._v1_scan_count = 0
+        self._v2_scan_count = 0
+        self._device_scan_counts = {}
```

Comment:
There is going to be a race here like #652 when you iterate over this dictionary since the dictionary size can change at any time from a background thread if a new device happens to be seen

Reply:
I should be a bit safer here, since I'm not doing a full iteration so much as a conditional. That is, the check for "does the key exist / add it, then +1 that key" should be sequentially uninterruptible by a call to reset_scan_stats(), since generally our GIL wouldn't likely interrupt.

However, I adjusted the implementation to use a single operation, which should be fully thread safe without needing a lock:
`self._device_scan_counts.setdefault(device_uuid, {'v1': 0, 'v2': 0})['v1'] += 1`

Reply:
Understood.  I was more worried about if you were going to return `_device_scan_counts` to someone outside of this class.  That's what we did in #652 and can result in the caller having issues if they try to iterate over the result you returned them, e.g. to print it out and the result changes size during their iteration because a new device is seen.  

I think that would still be an issue even with the new code because the caller can take as long as they want to iterate over `self._device_scan_counts`.  A fix if the intent of `self._device_scan_counts()` is to hand it over to an outside caller would be to hand them a copy generated by `dict.copy()`:
https://blog.labix.org/2008/06/27/watch-out-for-listdictkeys-in-python-3

Actually this also resolves #652...

Reply:
If we had an iterator in the class that wanted to go over self._device_scan_counts, we should make sure that iterator does copy in its implementation as you mentioned above. The returned dictionary does not appear to be impacted on iteration though, because it does create a copy. I tested this by grabbing the returned dictionary, and resetting it during an iteration over the items:
```
>>> (s, v1, v1r, v2, devs, timedelta) = g.device_manager.adapters[0].get_scan_stats()
>>> for dev in devs.items():
...  print(dev)
...  g.device_manager.adapters[0].reset_scan_stats()
... 
(2779, {'v1': 0, 'v2': 170})
(2783, {'v1': 583, 'v2': 0})
(2092, {'v1': 89, 'v2': 0})
(2780, {'v1': 0, 'v2': 167})
(442, {'v1': 196, 'v2': 0})
(263, {'v1': 75, 'v2': 0})
(1751, {'v1': 85, 'v2': 0})
(2778, {'v1': 0, 'v2': 165})
(1330, {'v1': 93, 'v2': 0})
(1877, {'v1': 12, 'v2': 0})
(2093, {'v1': 46, 'v2': 0})
(1443, {'v1': 38, 'v2': 0})
(1473, {'v1': 36, 'v2': 0})
(2215, {'v1': 79, 'v2': 0})
(1339, {'v1': 0, 'v2': 92})
(557, {'v1': 46, 'v2': 0})
(2095, {'v1': 100, 'v2': 0})
(2633, {'v1': 0, 'v2': 39})
(2098, {'v1': 10, 'v2': 0})
(414, {'v1': 6, 'v2': 0})
(1337, {'v1': 1, 'v2': 0})
(1875, {'v1': 1, 'v2': 0})
```